Add 64 bit atomics

atlv24 opened the pull request:
**Connections**
Part of an incoming series of PRs needed by #5123
Others in this series: #5155 #5154
Addresses #4424
Depends on #5498 and #5497
Ultimately for bevy meshlets pipeline https://github.com/bevyengine/bevy/pull/10164

**Description**
Shaders currently do not support 64 bit integer atomics. Add support for these gated by a feature flag when the capability is present.

Metal 64bit atomics are quite different from VK/DX style atomics, they do not have a return/output value of any kind and are only available in min/max variants. To accommodate this, a separate feature was added for min/max 64 bit atomics, and AtomicFunctionNoReturn was created in the naga IR to represent these return-less atomics.

DirectX ShaderModel version detection was also implemented, as SM 6.0 was insufficient for this feature.

**Testing**
A couple snapshot tests have been added to test int64 atomic operations.

<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

cwfitzgerald commented on wgpu-hal/src/dx12/adapter.rs:
```diff
@@ -311,6 +311,23 @@ impl super::Adapter {
+                    mem::size_of::<crate::dx12::types::D3D12_FEATURE_DATA_D3D12_OPTIONS9>() as _,
+                )
+            };
+            hr == 0 && features9.AtomicInt64OnTypedResourceSupported != 0
```
I don't think this check is right.

cwfitzgerald replied:
I'm asking for clarification on the features.

cwfitzgerald replied:
To quote Jesse:

> - Raw buffers are required to be supported by SM6.6 without any additional optional caps.
> - A typed resource would be a RWBuffer<sometype> or RWTexture2D<sometype>
> - A descriptor heap resource is one accessed via ResourceDescriptorHeap[index]

So if we just want buffer and group shared (which is what webgpu supports) we want to check only `AtomicInt64OnGroupSharedSupported`

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -351,6 +355,17 @@ impl PhysicalDeviceFeatures {
             } else {
                 None
             },
+            shader_int64_atomic: if device_api_version >= vk::API_VERSION_1_1
```
This seems to be checking if image atomics are supported, not if regular 64bit atomics are supported.

cwfitzgerald commented on naga/tests/in/atomicCompareExchange-int64.wgsl:
These shaders need to be regenerated.

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -1458,6 +1483,11 @@ impl super::Adapter {
             }
 
+            if features.contains(wgt::Features::SHADER_INT64_ATOMIC) {
+                capabilities.push(spv::Capability::Int64ImageEXT);
```
Same wrt images.

JMS55 commented on naga/src/valid/mod.rs:
```diff
@@ -110,6 +110,8 @@ bitflags::bitflags! {
         const CUBE_ARRAY_TEXTURES = 0x4000;
         /// Support for 64-bit signed and unsigned integers.
         const SHADER_INT64 = 0x8000;
+        /// Support for 64-bit signed and unsigned integers.
```
Copy pasted comment, should mention atomics :)

JMS55 commented on wgpu-types/src/lib.rs:
```diff
@@ -883,6 +883,15 @@ bitflags::bitflags! {
+        ///
+        /// Supported platforms:
+        /// - Vulkan
+        /// - DX12 (DXC only)
```
Should maybe mention the VK extension required, and the DXC shader model/feature/thing? Idk how widespread support is / how new this is.

atlv24 commented on CHANGELOG.md:
```diff
@@ -99,7 +99,8 @@ Bottom level categories:
   By @ErichDonGubler in [#5146](https://github.com/gfx-rs/wgpu/pull/5146), [#5046](https://github.com/gfx-rs/wgpu/pull/5046).
-- Signed and unsigned 64 bit integer support in shaders. By @rodolphito and @cwfitzgerald in [#5154](https://github.com/gfx-rs/wgpu/pull/5154)
+- Signed and unsigned 64 bit integer support in shaders. By @atlv24 and @cwfitzgerald in [#5154](https://github.com/gfx-rs/wgpu/pull/5154)
+- 64 bit integer atomic support in shaders. By @atlv24 in [#5383](https://github.com/gfx-rs/wgpu/pull/5383)
```
i've changed my github username from rodolphito to atlv24

cwfitzgerald commented on wgpu-core/src/device/resource.rs:
```diff
@@ -1525,6 +1525,16 @@ impl<A: HalApi> Device<A> {
+        caps.set(
+            Caps::SHADER_INT64_ATOMIC_ALL_OPS,
+            self.features
+                .contains(wgt::Features::SHADER_INT64_ATOMIC_ALL_OPS),
```
All on the name here feels weird considering you need the other feature to turn on Min/Max

atlv24 replied:
`SHADER_INT64_ATOMIC_ALL_OPS` does include min/max, though

atlv24 replied:
I made the code more explicit

cwfitzgerald reviewed (changes requested):
Looking good from the wgpu side, small comments.

There should be at least one execution test that makes sure everything at least seems to behave correctly and gets through all validation layers and such. Check calling atomicAdd N times results in a value of N, etc.

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -448,6 +454,17 @@ impl PhysicalDeviceFeatures {
+            {
+                Some(
+                    vk::PhysicalDeviceShaderAtomicInt64Features::builder()
+                        .shader_buffer_int64_atomics(true)
```
This is checking for VK_KHR_get_physical_device_properties2 not VK_KHR_shader_atomic_int64?

cwfitzgerald commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -109,6 +109,9 @@ pub struct PhysicalDeviceFeatures {
     zero_initialize_workgroup_memory:
         Option<vk::PhysicalDeviceZeroInitializeWorkgroupMemoryFeatures>,
 
+    /// Features provided by `VK_KHR_shader_atomic_int64`, promoted to Vulkan 1.3.
```
```suggestion
    /// Features provided by `VK_KHR_shader_atomic_int64`, promoted to Vulkan 1.2.
```

jimblandy commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -110,6 +110,9 @@ pub struct PhysicalDeviceFeatures {
         Option<vk::PhysicalDeviceZeroInitializeWorkgroupMemoryFeatures<'static>>,
 
+    /// Features provided by `VK_KHR_shader_atomic_int64`, promoted to Vulkan 1.2.
+    shader_int64_atomic: Option<vk::PhysicalDeviceShaderAtomicInt64Features<'static>>,
```
Please rename this `shader_atomic_int64`. I recognize that this matches the `wgpu_types::Features` name, but generally the `PhysicalDeviceFeatures` field for `vk::PhysicalDeviceMumbleFeatures` is named `mumble`, and I think that's the stronger precedent here.

jimblandy commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -423,6 +429,16 @@ impl PhysicalDeviceFeatures {
+            {
+                Some(
+                    vk::PhysicalDeviceShaderAtomicInt64Features::default()
+                        .shader_buffer_int64_atomics(true),
```
Shouldn't this consult `requested_features`, rather than hard-wiring `true`?

jimblandy commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -563,6 +579,11 @@ impl PhysicalDeviceFeatures {
+            F::SHADER_INT64_ATOMIC_ALL_OPS | F::SHADER_INT64_ATOMIC_MIN_MAX,
+            caps.supports_extension(khr::shader_atomic_int64::NAME),
+        );
+
```
I think this isn't correct. Simply because a Vulkan physical device supports the `VK_KHR_shader_atomic_int64` extension, it doesn't follow that it actually supports 64-bit atomic operations. The Vulkan design is such that the Mumble extension could be present, but then the `VkPhysicalDeviceMumbleFeature` fields could still be `false`.

I think you need to handle this more like `F::MULTIVIEW` later in this function: check whether `self.shader_int64_atomic` is `Some`, and if so, set the `wgpu_types::Features` according to the fields in `vk::PhysicalDeviceShaderAtomicInt64Features`.

jimblandy commented on naga/src/lib.rs:
```diff
@@ -1107,6 +1107,25 @@ pub enum AtomicFunction {
+    Max,
+}
+
+impl AtomicFunctionNoReturn {
+    pub const fn with_return(self) -> AtomicFunction {
+        match self {
+            Self::Min => AtomicFunction::Min,
+            Self::Max => AtomicFunction::Max,
+        }
+    }
+}
+
```
This function probably belongs in `naga/src/proc/mod.rs`, not in `lib.rs`.

jimblandy commented on naga/src/valid/mod.rs:
```diff
@@ -114,6 +114,10 @@ bitflags::bitflags! {
         const SUBGROUP = 0x10000;
         /// Support for subgroup barriers.
         const SUBGROUP_BARRIER = 0x20000;
+        /// Support for 64-bit integer atomics.
+        const SHADER_INT64_ATOMIC_MIN_MAX = 0x40000;
+        /// Support for 64-bit integer atomics.
+        const SHADER_INT64_ATOMIC_ALL_OPS = 0x80000;
```
These two comments aren't even different. Could you provide more detail?

jimblandy reviewed (changes requested):
Still reviewing, this is just the stuff I've noticed so far.

jimblandy commented on naga/src/lib.rs:
```diff
@@ -1900,6 +1919,15 @@ pub enum Statement {
         /// [`AtomicResult`]: crate::Expression::AtomicResult
         result: Handle<Expression>,
     },
+    /// Atomic function with no return value.
+    AtomicNoReturn {
+        /// Pointer to an atomic value.
+        pointer: Handle<Expression>,
+        /// Function to run on the atomic.
+        fun: AtomicFunctionNoReturn,
+        /// Value to use in the function.
+        value: Handle<Expression>,
+    },
```
Let's get some proper documentation on this. What does it do? What types does it support? What address spaces are permitted under which features? Why is it different from `Atomic`?

I've made one attempt at reading through the code in `valid/function.rs`, but I'm still struggling to understand exactly which behaviors this statement supports and does not support. Ideally the documentation in `lib.rs` should fully explain the behavior and constraints of each statement.

jimblandy reviewed (changes requested):
Before I review this further I'd like to see the doc comments filled in enough.

jimblandy commented:
@atlv24 Thanks very much for the new comments - these are helpful.

jimblandy commented:
I wanted to make some changes to this PR, but in the process came across a few things that need to be fixed first. #5735 was my misunderstanding, but I think #5741 and #5763 need to be addressed. Those are both in progress.

jimblandy commented:
@atlv24 I just pushed a new commit to this branch; the log message is below. Could you look it over and see what you think?

> Make `Statement::Atomic::result` optional.
> 
> Rather than introducing an entirely new statement variant `AtomicNoResult`, make `Statement::Atomic::result` an `Option`. This reduces the patch size [edit] a bit, and has no effect on snapshot output other than the direct IR dumps.
> 
> In HLSL output, don't bother to generated "discard" temporaries for functions where the final `original_value` parameter is not required. This improves snapshot output.
> 
> Improve documentation for `Expression::AtomicResult`, `Statement::Atomic`, and the new `valid::Capabilities` flags.
> 
> Rewrite `Atomic` statement validation. Consolidate validation of `AtomicResult` expressions with this; #5771 ensures that there is indeed some `Atomic` statement referring to every `AtomicResult` expression, so we can be sure it will be validated.
> 
> Fixes #5742.

jimblandy commented:
(No offense meant by removing your docs, after asking you to write them - they were helpful in getting started!)

jimblandy commented:
> but I think #5741 and #5763 need to be addressed.

#5741 is fixed. #5763 is not, but I think fixing #5742 should be good enough for now.

jimblandy commented:
Merged trunk so CI would pay attention

jimblandy reviewed (approved):
If my changes look good, then the rest of this looks good to me.

teoxoy commented on wgpu-hal/src/dx12/adapter.rs:
```diff
@@ -354,6 +354,23 @@ impl super::Adapter {
+                    mem::size_of::<crate::dx12::types::D3D12_FEATURE_DATA_D3D12_OPTIONS9>() as _,
+                )
+            };
+            hr == 0 && features9.AtomicInt64OnGroupSharedSupported != 0
```
Shouldn't we check for `AtomicInt64OnTypedResourceSupported` as well?

teoxoy commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -559,6 +578,13 @@ impl PhysicalDeviceFeatures {
+        if let Some(ref shader_atomic_int64) = self.shader_atomic_int64 {
+            features.set(
+                F::SHADER_INT64_ATOMIC_ALL_OPS | F::SHADER_INT64_ATOMIC_MIN_MAX,
+                shader_atomic_int64.shader_buffer_int64_atomics != 0,
```
Shouldn't we check for `shaderSharedInt64Atomics` as well?

jimblandy replied:
> - `shaderSharedInt64Atomics` indicates whether shaders can perform 64-bit unsigned and signed integer atomic operations on shared and payload memory.

Does "shared" mean WGSL `var<workgroup>`? What is "payload"?

/me searches the spec frantically

teoxoy replied:
> Does "shared" mean WGSL `var<workgroup>`?

That was my read, D3D12's `AtomicInt64OnGroupSharedSupported` is more explicit.

> `AtomicInt64OnGroupSharedSupported` is a boolean that specifies whether 64-bit integer atomics are supported on `groupshared` variables.

> What is "payload"?

I have no idea 😅

jimblandy replied:
I did not succeed in finding clearly relevant definitions of these terms.

teoxoy reviewed (commented):
Took a brief look at the feature detection code.

jimblandy commented:
> If my changes look good, then the rest of this looks good to me.

@atlv24 This is Just waiting for your comments on my commit, and for you to look over @teoxoy's review.

atlv24 commented:
@jimblandy I reviewed, looks great! I wish I had thought of using Option myself haha. Would have saved me a lot of copy pasting :)

atlv24 committed:
Add support for 64 bit integer atomic operations in shaders.

Add the following flags to `wgpu_types::Features`:

- `SHADER_INT64_ATOMIC_ALL_OPS` enables all atomic operations on `atomic<i64>` and
  `atomic<u64>` values.

- `SHADER_INT64_ATOMIC_MIN_MAX` is a subset of the above, enabling only
  `AtomicFunction::Min` and `AtomicFunction::Max` operations on `atomic<i64>` and
  `atomic<u64>` values in the `Storage` address space. These are the only 64-bit
  atomic operations available on Metal as of 3.1.

Add corresponding flags to `naga::valid::Capabilities`. These are supported by the
WGSL front end, and all Naga backends.

Platform support:

- On Direct3d 12, in `D3D12_FEATURE_DATA_D3D12_OPTIONS9`, if
  `AtomicInt64OnTypedResourceSupported` and `AtomicInt64OnGroupSharedSupported` are
  both available, then both wgpu features described above are available.

- On Metal, `SHADER_INT64_ATOMIC_MIN_MAX` is available on Apple9 hardware, and on
  hardware that advertises both Apple8 and Mac2 support. This also requires Metal
  Shading Language 2.4 or later. Metal does not yet support the more general
  `SHADER_INT64_ATOMIC_ALL_OPS`.

- On Vulkan, if the `VK_KHR_shader_atomic_int64` extension is available with both the
  `shader_buffer_int64_atomics` and `shader_shared_int64_atomics` features, then both
  wgpu features described above are available.
